File: JavaScript/src/Internals/Functions/Financial/Function_FV.js
import { Function_N } from '../Function_N.js';
import { Operand } from '../../../Operand.js';

class Function_FV extends Function_N {
    get Name() {
        return "FV";
    }

    constructor(funcs) {
        super(funcs);
    }

    evaluate(engine, tempParameter) {
        if (this.z.length < 3) return this.parameterError(1);

        const rateArg = this.getNumber(engine, tempParameter, 0);
        if (rateArg.IsError) return rateArg;
        const rate = rateArg.NumberValue;

        const nperArg = this.getNumber(engine, tempParameter, 1);
        if (nperArg.IsError) return nperArg;
        const nper = nperArg.NumberValue;

        const pmtArg = this.getNumber(engine, tempParameter, 2);
        if (pmtArg.IsError) return pmtArg;
        const pmt = pmtArg.NumberValue;

        let pv = 0;
        if (this.z.length > 3) {
            const pvArg = this.getNumber(engine, tempParameter, 3);
            if (pvArg.IsError) return pvArg;
            pv = pvArg.NumberValue;
        }

        let type = 0;
        if (this.z.length > 4) {
            const typeArg = this.getNumber(engine, tempParameter, 4);
            if (typeArg.IsError) return typeArg;
            type = typeArg.IntValue;
            if (type !== 0 && type !== 1) return this.parameterError(5);
        }

        if (rate === 0) {
            return Operand.Create(-pv - pmt * nper);
        }

        const factor = Math.pow(1 + rate, nper);
        let fv;
        if (type === 1) {
            // 期初支付：仅 pmt 项乘 (1+rate)，与 C#/Excel 一致
            fv = -pv * factor - pmt * (1 + rate) * (factor - 1) / rate;
        } else {
            fv = -pv * factor - pmt * (factor - 1) / rate;
        }

        return Operand.Create(fv);
    }
}

export { Function_FV };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/financial/Function_FV.java
package toolgood.algorithm.internals.functions.financial;

import java.math.BigDecimal;
import java.math.MathContext;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_5;

public final class Function_FV extends Function_5 {
    public Function_FV(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length < 3 || funcs.length > 5) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires 3 to 5 parameters.");
        }
    }

    @Override
    public String Name() {
        return "FV";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand rateArg = GetNumber_1(engine, tempParameter);
        if (rateArg.IsErrorOrNone()) return rateArg;
        BigDecimal rate = rateArg.NumberValue();

        Operand nperArg = GetNumber_2(engine, tempParameter);
        if (nperArg.IsErrorOrNone()) return nperArg;
        BigDecimal nper = nperArg.NumberValue();

        Operand pmtArg = GetNumber_3(engine, tempParameter);
        if (pmtArg.IsErrorOrNone()) return pmtArg;
        BigDecimal pmt = pmtArg.NumberValue();

        BigDecimal pv = BigDecimal.ZERO;
        if (func4 != null) {
            Operand pvArg = GetNumber_4(engine, tempParameter);
            if (pvArg.IsErrorOrNone()) return pvArg;
            pv = pvArg.NumberValue();
        }

        int type = 0;
        if (func5 != null) {
            Operand typeArg = GetNumber_5(engine, tempParameter);
            if (typeArg.IsErrorOrNone()) return typeArg;
            type = typeArg.IntValue();
            if (type != 0 && type != 1) {
                return ParameterError(5);
            }
        }

        if (rate.compareTo(BigDecimal.ZERO) == 0) {
            return Operand.Create(pmt.multiply(nper).negate().subtract(pv));
        }

        BigDecimal factor = BigDecimal.valueOf(Math.pow(BigDecimal.ONE.add(rate).doubleValue(), nper.doubleValue()));
        BigDecimal fv = pv.multiply(factor).negate()
                .subtract(pmt.multiply(factor.subtract(BigDecimal.ONE)).divide(rate, MathContext.DECIMAL128));
        if (type == 1) {
            fv = pv.multiply(factor).negate()
                    .subtract(pmt.multiply(BigDecimal.ONE.add(rate)).multiply(factor.subtract(BigDecimal.ONE)).divide(rate, MathContext.DECIMAL128));
        }

        return Operand.Create(fv);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func3.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        if (func4 != null) func4.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        if (func5 != null) func5.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}
